mipengine/algorithms/ttest_independent.py

```diff
@@ -0,0 +1,180 @@
+import json
+
+import numpy
+from pydantic import BaseModel
+
+from mipengine.udfgen import secure_transfer
+from mipengine.udfgen.udfgenerator import literal
+from mipengine.udfgen.udfgenerator import relation
+from mipengine.udfgen.udfgenerator import transfer
+from mipengine.udfgen.udfgenerator import udf
+
+
+class TtestResult(BaseModel):
+    t_stat: float
+    df: int
+    p: float
+    mean_diff: float
+    se_diff: float
+    ci_upper: float
+    ci_lower: float
+    cohens_d: float
+
+
+def run(algo_interface):
+    local_run = algo_interface.run_udf_on_local_nodes
+    global_run = algo_interface.run_udf_on_global_node
+    alpha = algo_interface.algorithm_parameters["alpha"]
+    alternative = algo_interface.algorithm_parameters["alt_hypothesis"]
+
+    X_relation, Y_relation = algo_interface.create_primary_data_views(
+        variable_groups=[algo_interface.x_variables, algo_interface.y_variables],
+    )
+
+    sec_local_transfer = local_run(
+        func=local_independent,
+        keyword_args=dict(y=Y_relation, x=X_relation),
+        share_to_global=[True],
+    )
+
+    result = global_run(
+        func=global_independent,
+        keyword_args=dict(
+            sec_local_transfer=sec_local_transfer, alpha=alpha, alternative=alternative
+        ),
+    )
+
+    result = json.loads(result.get_table_data()[1][0])
+    res = TtestResult(
+        t_stat=result["t_stat"],
+        df=result["df"],
+        p=result["p"],
+        mean_diff=result["mean_diff"],
+        se_diff=result["se_diff"],
+        ci_upper=result["ci_upper"],
+        ci_lower=result["ci_lower"],
+        cohens_d=result["cohens_d"],
+    )
+
+    return res
+
+
+@udf(
+    y=relation(),
+    x=relation(),
+    return_type=[secure_transfer(sum_op=True)],
+)
+def local_independent(x, y):
+    x1 = x.reset_index(drop=True).to_numpy().squeeze()
+    x2 = y.reset_index(drop=True).to_numpy().squeeze()
+    x1_sum = sum(x1)
+    x2_sum = sum(x2)
+    n_obs_x1 = len(x)
+    n_obs_x2 = len(y)
+    diff = sum(x1 - x2)
+    diff_sqrd = sum((x1 - x2) ** 2)
+    x1_sqrd_sum = sum(x1**2)
+    x2_sqrd_sum = sum(x2**2)
+
+    sec_transfer_ = {
+        "n_obs_x1": {"data": n_obs_x1, "operation": "sum", "type": "int"},
+        "n_obs_x2": {"data": n_obs_x2, "operation": "sum", "type": "int"},
+        "sum_x1": {"data": x1_sum.item(), "operation": "sum", "type": "float"},
+        "sum_x2": {"data": x2_sum.item(), "operation": "sum", "type": "float"},
+        "diff": {"data": diff.tolist(), "operation": "sum", "type": "float"},
+        "diff_sqrd": {"data": diff_sqrd.tolist(), "operation": "sum", "type": "float"},
+        "x1_sqrd_sum": {
+            "data": x1_sqrd_sum.tolist(),
+            "operation": "sum",
+            "type": "float",
+        },
+        "x2_sqrd_sum": {
+            "data": x2_sqrd_sum.tolist(),
+            "operation": "sum",
+            "type": "float",
+        },
+    }
```

Comment:
You don't need to send `sum_x1`, `sum_x2` and `diff` since the latter is calculated from the first two. The same applies for `x1_sqrd_sum`, `x2_sqrd_sum` and `diff_sqrd`.